### PR TITLE
rewrite

### LINKED ISSUE #6: Misplaced "Reset locale" button in menu
See screenshot, "Reset locale" button extends beyond the right border of drop-down menu.
openSUSE Leap 15.1
Chrome 80.0.3987.87 obtained from Google repository.
Addon v. 0.1.1

Screenshot_20200210_192444

### LINKED ISSUE #8: Allow changing locale only for active tab
From user request:

image

### LINKED ISSUE #23: "Reset" doesn't fit in box
20211130T064906

Also is this app maintained anymore?

### LINKED ISSUE #25: Add shortcuts for easy interaction
It would be cool to have a shortcut to open the options, so everything can be done with a keyboard

### PR DESCRIPTION
**Pull request by cupcakearmy**
- fixes #6 
- fixes #23 
- closes #8 
- closes #25

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Changelog
```
Comment (athyuttamre) [inline CHANGELOG.md:1]: Is this changelog for this project? Seems unrelated, maybe we can drop this file?

File: README.md [hunk 2]
```diff
@@ -1,3 +1,49 @@
-# Locale Switcher
```
Comment (athyuttamre) [inline README.md:1]: This readme seems to be for another project. Can we revert?

File: manifest.json [hunk 3]
```diff
@@ -1,19 +1,18 @@
...
   "name": "Locale Switcher",
-  "version": "0.1.1",
+  "version": "1.0.0",
   "description": "Lets you quickly switch the browser locale to test localization on your website.",
   "permissions": ["storage", "webRequest", "webRequestBlocking", "<all_urls>"],
   "minimum_chrome_version": "72",
   "background": {
-    "scripts": ["background.js"],
-    "persistent": true
+    "scripts": ["./src/background/index.ts"]
```
Comment (athyuttamre) [inline manifest.json:9]: The webextension-polyfill docs say we need to add a polyfill script here; do we not need to do that? 

https://github.com/mozilla/webextension-polyfill#basic-setup

File: pnpm-lock.yaml [hunk 4]
```diff
@@ -0,0 +1,6816 @@
+lockfileVersion: 5.3
```
Comment (athyuttamre) [inline pnpm-lock.yaml:1]: Do we need this file? Can we use vanilla npm?

File: src/background/index.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
+import browser, { WebRequest } from 'webextension-polyfill'
+import type { Locale, MessageType } from '../types'
+
+declare const chrome: any
```
Comment (athyuttamre) [inline src/background/index.ts]: I wonder if this package can help with Chrome types: https://www.npmjs.com/package/@types/chrome

File: src/background/index.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+let locale: Locale = null
+
+function updateLocale(value: Locale) {
+  if (value === locale) return
+  locale = value || null
+  browser.storage.local.set({ locale })
+  browser.browserAction.setBadgeText({ text: locale })
+}
+
+async function sendMessageToActiveTabs(message: MessageType) {
```
Comment (athyuttamre) [inline src/background/index.ts]: Since we're filtering by current window, I assume this only sends a message to *one* tab?

File: src/popup/lib/store.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import { writable } from 'svelte/store'
+import browser from 'webextension-polyfill'
+import type { Locale } from '../../types'
+
+export const locale = writable<Locale>(null)
+
+browser.storage.local.get('locale').then((result) => {
```
Comment (athyuttamre) [inline src/popup/lib/store.ts]: What happens if the popup is open but the tab changes? How does the popup get notified of locale updates?

File: src/content/index.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+const KEY = 'LOCALE_SWITCHER_LANGUAGE'
+
+function embedScript() {
+  const locale = window.localStorage.getItem(KEY)
+  if (!locale) return
+
+  const code = `
+    (() => {
+      Object.defineProperties(window.navigator.__proto__, {
```
Comment (athyuttamre) [inline src/content/index.ts:12]: I previously used `Navigator.prototype` here. Is there a difference between that and this?

File: src/background/index.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      }
+    }
+    return { requestHeaders: details.requestHeaders || [] }
+  },
+  { urls: ['<all_urls>'] },
+  options
+)
+
+// Update on tab change or new urls
+browser.tabs.onActivated.addListener(() => getCurrentLocale())
```
Comment (athyuttamre) [inline src/background/index.ts]: There's a subtle race condition here: when the active tab changes, we're asynchronously fetching the tab's locale. However, in the meantime the tab might have made network requests, and those will not pick up the right locale. I think this is also solved by maintaining a mapping from tabId to locale in the background script.

File: src/content/index.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  const script = document.createElement('script')
+  script.textContent = code
+  document.documentElement.prepend(script)
+  script.remove()
+}
+
+function handleMessage({ type, data }: MessageType) {
+  switch (type) {
+    case 'update':
+      if (data === window.localStorage.getItem(KEY)) break
```
Comment (athyuttamre) [inline src/content/index.ts]: In my testing in Chrome `window.localStorage` was not accessible inside extension event listeners. This is why I had to use the same injection trick above to change the localStorage / sessionStorage. Did this work for you in Chrome?

https://github.com/locale-switcher/locale-switcher/commit/d7ace06e091e91d179cb4e5b9c770e9d42d6820e#diff-c89f55ed43133253050dbce66bfd2c8a4f6280615dbaa582b7efbe3104a79614R17

**Review by athyuttamre — COMMENTED** [review]
Thanks for doing this! 

I think the dataflow has a few issues:

- Because we're using localStorage, the locale persists across all tabs of the same domain. Some users may want each tab to be different, especially if they want to compare their website in two different locales side by side. So I think we should use sessionStorage.

- Because we store only one locale at a time in the background script, any network requests made by inactive tabs (i.e. not the current tab) will use the current tab's locale instead. This is the wrong behavior, so I think we need to store a mapping between tab IDs and locales in the background script. See my branch for an example: https://github.com/locale-switcher/locale-switcher/commit/d7ace06e091e91d179cb4e5b9c770e9d42d6820e#diff-bb06783fccd762e4d905ab922491e954a9fbaf708a1acbf53dff92872b0bcd3bR5.

- It seems like we have two persistent storage solutions here: the tab's localStorage/sessionStorage, and the extension's `browser.storage.local` API. Do we need both? I don't have a specific issue in mind, but in general having one source of persistent truth is better than having two, and the tab's localStorage / sessionStorage seems like the better one. The popup can get/update the locale by sending messages instead of setting/getting the `browser.storage.local` value. I'm not super sure about this one, so open to your thoughts.

Some more questions inline!

File: README.md (hunk 2 above)
Comment (cupcakearmy) [reply README.md:1]: Jup 🤦‍♂️

File: manifest.json (hunk 3 above)
Comment (cupcakearmy) [reply manifest.json:9]: The polyfill is made by `import browser from 'webextension-polyfill'`

File: pnpm-lock.yaml (hunk 4 above)
Comment (cupcakearmy) [reply pnpm-lock.yaml:1]: Mhh i really like pnpm, is way faster and uses a fraction of disk space.

File: src/background/index.ts (hunk 5 above)
Comment (cupcakearmy) [reply src/background/index.ts]: good idea, just checked but they seem to not be good enough, they don't have `OnBeforeSendHeadersOptions`

File: src/background/index.ts (hunk 6 above)
Comment (cupcakearmy) [reply src/background/index.ts]: Basically yes, its only one tab but wanted to keep it generic

File: src/content/index.ts (hunk 8 above)
Comment (cupcakearmy) [reply src/content/index.ts:12]: Mhh not really, but since most libs take `window.navigator` as the source I thought it might be a good idea to use that. Both seem to work, however they are not the same object according to `===`. Quite werid, but works

File: src/popup/lib/store.ts (hunk 7 above)
Comment (cupcakearmy) [reply src/popup/lib/store.ts]: I though about that, but since the popup is the only one that can change that it should not be an issue.
Nothing updates the locale outside of the popup

File: src/background/index.ts (hunk 9 above)
Comment (cupcakearmy) [reply src/background/index.ts]: Mhh the race condition should be fine, as that is only to pass it to the popup. The locale that is taken is saved locally.

File: src/background/index.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  const active = await browser.tabs.query({ active: true, currentWindow: true })
+  for (const tab of active) {
+    if (tab.id) browser.tabs.sendMessage(tab.id, message)
+  }
+}
+
+async function getCurrentLocale() {
+  const active = await browser.tabs.query({ active: true, currentWindow: true })
+  for (const tab of active) {
+    if (tab.id) browser.tabs.sendMessage(tab.id, { type: 'get' })
```
Comment (athyuttamre) [inline src/background/index.ts]: On testing locally, there seems to be another bug:

1. Download the extension and install it
2. Set the locale on one page, say "Tab A"
3. Switch to another page that was already open when the extension was installed, say "Tab B"
4. Notice that the popup's badge and locale value do not set to null as expected. This is because the content script is not yet initialized on "Tab B" so we never get a response for this message.

File: src/background/index.ts (hunk 9 above)
Comment (cupcakearmy) [reply src/background/index.ts]: Nvm, I see what you mean. However the tab gets reloaded when it's changed. We will always have some kind of race condition as the commnunication is async.
Also generally it should be fine as the switch goes from: popup->bg->client.
Assuming you have multiple tabs, switchig and reloading faster that the message, but seems unlikely.

File: src/popup/components/Footer.svelte [hunk 12]
```diff
@@ -0,0 +1,4 @@
+<div class="tr code f7">
+  <a href="https://github.com" target="_blank" rel="noreferrer">Github</a>
+  ﹣ v1.1.0
```
Comment (athyuttamre) [inline src/popup/components/Footer.svelte]: We should update the link and version number here, or we could also drop the footer altogether since it's easy to forget to update this.

File: src/content/index.ts (hunk 10 above)
Comment (cupcakearmy) [reply src/content/index.ts]: Mh.. it works for me in chrome?

File: src/background/index.ts (hunk 11 above)
Comment (cupcakearmy) [reply src/background/index.ts]: Done

**Comment by cupcakearmy** [comment]
Sooo, locales are now session based, tab scoped with full sync back to the popup. (you can even switch tabs and the popup will update while being open)

File: src/background/index.ts [hunk 13]
```diff
@@ -46,23 +44,40 @@ browser.webRequest.onBeforeSendHeaders.addListener(
...
+  browser.runtime.sendMessage(msg)
+}
 
 // Listener for messages from content scripts
-browser.runtime.onMessage.addListener(({ type, data }) => {
-  switch (type) {
-    case 'get':
-      updateLocale(data)
+browser.runtime.onMessage.addListener(async (message: MessageType, sender) => {
+  switch (message.type) {
```
Comment (athyuttamre) [inline src/background/index.ts:61]: Could we name these events differently? Right now they're very generic words and a bit hard to inuitively understand what each one does.

Suggestions:
- `sync` -> `setBackgroundLocaleFromTab`
- `update` -> `setBackgroundLocaleFromPopup`
- `current-input` -> `getBackgroundLocale`
- `current-output` -> `setPopupLocale`

File: src/background/index.ts [hunk 14]
```diff
@@ -46,23 +44,40 @@ browser.webRequest.onBeforeSendHeaders.addListener(
...
+        updateLocale(tabId, message.data)
+        const msg: MessageType = { type: 'update', data: message.data }
+        browser.tabs.sendMessage(tabId, msg)
+      }
+      break
+    }
+
+    case 'current-input': {
+      const tabId = await getCurrentTabId()
+      if (tabId) sendCurrentState(tabId)
```
Comment (athyuttamre) [inline src/background/index.ts:80]: I think we can actually respond immediately using `sendResponse`, rather than broadcast another event: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage#sending_a_synchronous_response.

File: src/content/index.ts [hunk 15]
```diff
@@ -27,21 +27,24 @@ function embedScript() {
...
-      const locale = window.localStorage.getItem(KEY) || null
-      browser.runtime.sendMessage({ type: 'get', data: locale })
-      break
   }
 }
 
 browser.runtime.onMessage.addListener((message) => handleMessage(message))
+
 embedScript()
+sendCurrentState()
```
Comment (athyuttamre) [inline src/content/index.ts:50]: Why do we need to send a `sync` message on page load?

File: manifest.json [hunk 16]
```diff
@@ -28,5 +28,12 @@
     "32": "icons/32.png",
     "48": "icons/48.png",
     "128": "icons/128.png"
+  },
+  "commands": {
+    "_execute_browser_action": {
+      "suggested_key": {
+        "default": "Ctrl+Shift+L"
```
Comment (athyuttamre) [inline manifest.json:30]: Nice!

File: README.md [hunk 17]
```diff
@@ -1,3 +1,56 @@
...
+
+## Acknowledgements
+
+The data for the locales is scraped automatically from [Locale Plante](https://www.localeplanet.com/icu/index.html).
+
+## License
+
+Distributed under the MIT License. See `LICENSE` for more information.
+
+### Software used in this project
```
Comment (athyuttamre) [inline README.md]: Do we need this section? Can we drop it?

File: README.md [hunk 18]
```diff
@@ -1,3 +1,56 @@
...
+- Access to tabs to load the client script in order to change `navigator` properties.
+
+## Authors
+
+- Initial idea, implementation and distribution by @athyuttamre
+- New UI & current maintenance by @cupcakearmy
+
+## Acknowledgements
+
+The data for the locales is scraped automatically from [Locale Plante](https://www.localeplanet.com/icu/index.html).
```
Comment (athyuttamre) [inline README.md]: Typo: Locale Planet

File: README.md [hunk 19]
```diff
@@ -1,3 +1,56 @@
...
+- `navigator.language`
+- `navigator.languages`
+- `Accept-Language` headers
+
+## ✨ Features
+
+- Chrome & Firefox
+- Support for 700+ languages and variations
+- Per tab basis
+- Favorize your most used locales
```
Comment (athyuttamre) [inline README.md]: Favorite

File: README.md [hunk 20]
```diff
@@ -1,3 +1,56 @@
...
+- Fully shortcut enabled, no mouse needed
```
Comment (athyuttamre) [inline README.md]: Maybe mention the keyboard shortcut directly: `Keyboard shortcut `Ctrl+Shift+L` (Windows) or `Cmd+Shift+L` (Mac)`

File: README.md [hunk 21]
```diff
@@ -1,3 +1,56 @@
...
+
+- `navigator.language`
+- `navigator.languages`
+- `Accept-Language` headers
+
+## ✨ Features
+
+- Chrome & Firefox
+- Support for 700+ languages and variations
+- Per tab basis
```
Comment (athyuttamre) [inline README.md:21]: Perhaps: Changes the locale for the current tab only

File: README.md [hunk 22]
```diff
@@ -1,3 +1,56 @@
-# Locale Switcher
+<p align="center">
+  <img src="./icons/128.png" alt="logo" />
```
Comment (athyuttamre) [inline README.md]: I think we can remove this old logo from the README

File: README.md [hunk 23]
```diff
@@ -34,7 +34,7 @@ When a locale chosen the following thinks will be adjusted:
 
 ## Acknowledgements
 
-The data for the locales is scraped automatically from [Locale Plante](https://www.localeplanet.com/icu/index.html).
+The data for the locales is scraped automatically from [LocalePlantet](https://www.localeplanet.com/icu/index.html).
```
Comment (athyuttamre) [inline README.md]: Still a typo, should be "Locale Planet"

**Comment by athyuttamre** [comment]
Thanks for the great work! I tested on both Chrome and Firefox, and they both work really well. Thank you!

File: README.md (hunk 21 above)
Comment (cupcakearmy) [reply README.md:21]: But you can change multiple tabs and it persists also, so tecnically it's scoped on the tab and not the current. Then again it might be easier to understadnt

File: README.md (hunk 22 above)
Comment (cupcakearmy) [reply README.md]: Its the new one :)

File: src/background/index.ts (hunk 13 above)
Comment (cupcakearmy) [reply src/background/index.ts:61]: Yeah, was strugling with renaming them and then forgot to do that xD

File: src/background/index.ts (hunk 14 above)
Comment (cupcakearmy) [reply src/background/index.ts:80]: True, but we needed an event anyways since we need to listen for changes, so I thought i'll keep it consistent

File: src/content/index.ts (hunk 15 above)
Comment (cupcakearmy) [reply src/content/index.ts:50]: It's not sync actually, the function is a non blocking promise
